Prototype Python layer for RInstatClimatic R package

lloyddewit opened the pull request:
Added a module that provides access to the `cdms.products` R package (https://github.com/IDEMSInternational/cdms.products). It communicates with the R environment using the `rpy2` package.
Access is provided through a set of wrapper functions.

lloyddewit committed:
Removed incorrect git folder from path

lloyddewit committed:
Reverted changes

lloyddewit committed:
naflex_na_omit_ifTestsAllPass

lloyddewit committed:
added signatures for 2 RInstatClimatic functions

lloyddewit committed:
implemented test_climatic_summary, passes

lloyddewit committed:
created tidy files for cdms_products

lloyddewit committed:
Added extra tests for climatic_summary

lloyddewit committed:
added timeseries plot fn and test

lloyddewit committed:
moved to cdms.products package dev version

lloyddewit committed:
refactored climatic_summary tests

lloyddewit committed:
inventory_table all tests pass

lloyddewit committed:
added inventory_plot all tests pass

lloyddewit committed:
added climsoft data and tests

lloyddewit committed:
added wrapper stubs, all tests pass

lloyddewit committed:
added types to wrapper functions

lloyddewit committed:
Added bodies to wrapper functions

lloyddewit committed:
updated tests

lloyddewit committed:
all tests pass except test_export_ tests

lloyddewit committed:
Removed unused files

lloyddewit committed:
added test placeholder for export_cdt

lloyddewit committed:
tidied test utility function

lloyddewit committed:
commented out test prior to delet

lloyddewit committed:
deleted test with problematic csv test data file

lloyddewit committed:
Refactoring and example Doc String

lloyddewit committed:
resolved data type todos

lloyddewit committed:
Resolved parameters with NA values

lloyddewit committed:
made file paths mandatory, all tests pass

lloyddewit committed:
added export_cdt test

lloyddewit committed:
Refactored imports

lloyddewit committed:
Added docstring for climatic_missing()

lloyddewit committed:
docstrings and timezone experiments

lloyddewit committed:
improved docstrings up to export_cdt()

lloyddewit committed:
improved docstrings, ongoing

lloyddewit committed:
Completed docstrings

lloyddewit committed:
completed docstrings for private functions

lloyddewit committed:
If dates don't include timezone, then set to UT

lloyddewit committed:
added new export tests

lloyddewit committed:
Added extra export tests, all tests pass

lloyddewit committed:
Updated histogram_plot warning test,all tests pass

lloyddewit committed:
Removed args and kwargs parameters

lloyddewit committed:
Removed logging notes

lloyddewit committed:
Fixed timestamp issue in test_inventory_plot

lloyddewit committed:
set dates to UTC, cleaned niger50 data

lloyddewit committed:
Added license header

lloyddewit committed:
Added module headers

lloyddewit committed:
Added development notes for cdms_products.py

lloyddewit commented:
@geoo89 As discussed a few weeks ago, please would it be possible to review this PR?

This PR has may files but they are mainly data files for testing. The only 3 files you really need to review are:

![image](https://user-images.githubusercontent.com/57253949/179459805-0bbab497-2ce6-4ea0-82ea-5092113c2742.png)

Thank you for your help with this! I'm happy to talk whenever needed

geoo89 commented on opencdms_process/process/rinstat/README.md:
````diff
@@ -0,0 +1,171 @@
+Check we have the latest version of Git, and set user name, email and credential manager. See [here](https://docs.microsoft.com/en-us/windows/wsl/tutorials/wsl-git).
+```bash
+sudo apt-get install git
+git config --global user.name "Stephen Lloyd"
+git config --global user.email stephen.lloyd03@gmail.com
````
I assume you do not necessarily want to put your e-mail address here, especially in case people copy/paste blindly?

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+sudo add-apt-repository 'deb https://cloud.r-project.org/bin/linux/ubuntu focal-cran40/'
+sudo apt update
+sudo apt install r-base
+sudo -i R
```
This opens the interactive R shell, but do you do anything with it afterwards? If not, I assume the last line is not needed (I see it used later).

geoo89 commented on opencdms_process/process/rinstat/README.md:
````diff
@@ -0,0 +1,171 @@
+sudo apt install r-base
+sudo -i R
+```
+### Fork GitHub repos
````
I assume forking is only necessary if you want to contribute to the repositories. Otherwise, the cloning instructions below should be sufficient.

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+# clone GitHub repos
+mkdir -p ~/opencdms/processes
+cd ~/opencdms/processes
+git clone https://github.com/lloyddewit/opencdms-process.git
```
`lloyddewit` should be replaced with the developer's personal fork.

geoo89 replied:
Also, from what I understand, the other two repositories are installed as dependencies automatically via the `requirements.txt`, so there's no need to fork or clone these.

geoo89 commented on opencdms_process/process/rinstat/README.md:
````diff
@@ -0,0 +1,171 @@
+For GitHub I installed manually.
+
+### Installed `cdms.products`
+```r
````
This this line (`r`) needed? The next line starts the interactive R console.

geoo89 commented on opencdms_process/process/rinstat/README.md:
````diff
@@ -0,0 +1,171 @@
+### Installed `cdms.products`
+```r
+sudo -i R
+devtools::install_github("IDEMSInternational/cdms.products")
````
I believe we need to do `install.packages('devtools')` before we can run this line.

geoo89 replied:
In order to install `devtools`, I had to install a couple of system libraries:
`sudo apt install libharfbuzz-dev libfribidi-dev libfreetype6-dev libpng-dev libtiff5-dev libjpeg-dev`. This may be different on Ubuntu 20.04, I don't know.

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+For Python, I was prompted automatically.
+For GitHub I installed manually.
+
+### Installed `cdms.products`
```
I would put this into the "Installing R and R libraries" section further up.

geoo89 commented on opencdms_process/process/rinstat/README.md:
````diff
@@ -0,0 +1,171 @@
+```r
+sudo -i R
+devtools::install_github("IDEMSInternational/cdms.products")
+is_available <- require("cdms.products")
````
This is about checking whether `cdms.products` has been successfully installed.
I would make a remark about this, and mention that the `TRUE` indicates that everything worked as intended.

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+I do not claim these notes are complete, or that this is the optimal development environment. I share these notes in case they are useful for other developers. Please feel free to correct or enhance these notes.
+
+We could also add a Wiki page for this repo.
+## Set up development environment
```
I would restructure this.
1. Minimal set up: this covers installation, being able to run the tests (and maybe a minimal working example of doing something basic with this library), verifying that everything was set up correctly.
    a. Windows specific setup (wsl, reference to installing git)
    b. Common setup instructions (the rest: system libraries needed for the R stuff, R, R libraries, system libraries needed for the python stuff, python, python virtual environment, python libraries)
3. Optional QOL components: VSCode, black, Jupyter Notebook.
4. Troubleshooting (i.e. the odd problems you had)
5. Appendix: Installing git, cloning/forking repos, etc. This is common knowledge for most developers, so I think it doesn't need to be featured in the main instructions. This could be referred to in the wsl setup as a necessary step.

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+# Development notes for `cdms_products.py`
```
What does this library do? Some intro would be good here.
What does `setup.py` do (not necessarily to be answered in the intro, but somewhere in the README)?

geoo89 commented on opencdms_process/process/rinstat/README.md:
````diff
@@ -0,0 +1,171 @@
+The RInstatClimatic installation failed trying to install the R `clifro` package with ‘Error, nc-config not found or not executable’.
+I found a solution [here](https://stackoverflow.com/questions/42891050/install-ncdf4-package-error-nc-config-not-found-or-not-executable).
+```bash
+sudo apt-get install libnetcdf-dev
````
I believe this line should also become part of the standard R installation instructions further up.

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+sudo add-apt-repository 'deb https://cloud.r-project.org/bin/linux/ubuntu focal-cran40/'
+sudo apt update
+sudo apt install r-base
+sudo apt install build-essential libcurl4-gnutls-dev libxml2-dev libssl-dev
```
This line should also become part of the default R/R libs installation instructions. (The duplicates in the python section can then be removed)

geoo89 commented on opencdms_process/process/rinstat/README.md:
````diff
@@ -0,0 +1,171 @@
+```bash
+#setup virtual environment
+sudo apt-get install python3-venv
+python3 -m venv opencdms-env
````
Add `opencdms-env/` to `.gitignore` so it doesn't appear as untracked file, or (better) rename this folder from `opencdms-env` to `env`, which seems to be more common practice as it's already included in the `.gitignore` that you already have here (the default one for python that github provides).

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+sudo apt install python3-pip
+pip3 install --upgrade pip
+
+# install mysqlclient, see https://askubuntu.com/questions/1321141/unable-to-install-mysqlclient-on-ubuntu-20-10
```
In these 3 blocks, it'd be good to understand which dependencies are needed for what.
- Who needs the mysqlclient? Instead of manually running `pip3 install mysqlclient>=2.0.3`, `mysqlclient>=2.0.3` should be in the relevant `requirements.txt`. Same for `psycopg2>=2.9.1` and `click>=7.1.2` (I see `click` is already in the requirements, so just add the version if it is necessary).
- `requirements.txt` for this project already contains the other two libraries as dependencies, which will in turn (I believe) recursively install their dependencies from their respective `requirements.txt`, so manually installing their dependencies shouldn't be needed. The exception may be `pyopencdms/requirements_dev.txt`, however, I was able to at least run the tests from this PR without doing `pip3 install -r ~/opencdms/processes/pyopencdms/requirements_dev.txt`.

geoo89 commented on opencdms_process/process/rinstat/README.md:
```diff
@@ -0,0 +1,171 @@
+sudo apt-get install -y python3-psycopg2
+sudo apt install libpq-dev
+pip3 install psycopg2>=2.9.1
+pip3 install -r ~/opencdms/processes/opencdms-process/requirements.txt
```
You're assuming a fixed installation path here (in accordance with your git instructions).
Better to use a relative path, i.e. `pip3 install -r requirements.txt`

geoo89 reviewed (changes requested):
I managed to get everything set up and run the tests natively on Ubuntu 18.04, however, most tests are failing. Can discuss this in a call.

The setup instructions are a bit unintuitive, so I've commented on the `README.md` for now. I will review the code and tests some other time.

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+    #    latitude = "lat", longitude = "long",
+    #    file_path = "C:\\Users\\steph\\OneDrive\\Desktop\\FirefoxDownloads\\export_geoclim_expected010.csv")
+    output_file_actual: str = os.path.join(
+        TEST_DIR, "results_actual", "export_geoclim_actual010"
```
`.csv` is missing here.

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+    # export_climdex(data = daily_niger, date = "date", prcp = "rain", tmax = "tmax", tmin = "tmin",
+    #       file_path = "C:\\Users\\steph\\OneDrive\\Desktop\\FirefoxDownloads\\export_climdex_expected010")
+    output_file_actual: str = os.path.join(
+        TEST_DIR, "results_actual", "export_climdex_actual010"
```
`.csv` seems to be missing here.

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+    #            element = "sum", latitude = "lat", longitude = "long",
+    #            altitude = "alt", type =  "daily", date_time = "date",
+    #            metadata = stations_niger,
+    #            file_path = "C:\\Users\\steph\\OneDrive\\Desktop\\FirefoxDownloads\\export_cdt_expected010.csv")
```
Remove the commented out lines. (There are more instances below which I'm not going to mention explicitly.)
I don't see the purpose they serve (it doesn't look like a complete R code to produce the expected output).

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+                     elements = c("tmin", "tmax"), facet_by = "elements")
+
+#inventory_plot
+mydata <- read.csv("C:\\Users\\steph\\OneDrive\\Desktop\\FirefoxDownloads\\observationFinalMinimal.csv")
```
Because of the hard-coded file path this isn't reproducible. Where do we find this file? Do we need it?

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+
+
+def test_climatic_extremes():
+    data_file: str = os.path.join(TEST_DIR, "data", "niger50.csv")
```
nit: slightly strange declaring the type here (in this mundane case) if you don't do it for anything else (also following instances).

geoo89 commented on opencdms_process/process/rinstat/cdms_products.py:
```diff
@@ -0,0 +1,1513 @@
+            if len(r_params[key]) > 0:
+                if isinstance(r_params[key][0], str):
+                    r_params[key] = StrVector(r_params[key])
+                elif isinstance(r_params[key][0], float):
```
The type annotations that you provide when defining functions are not necessarily respected (that's python for you). We should probably work out a robust way of inferring the type from the function annotations rather than the data itself.

If I pass a list like [1,2,3] (or even [1,2.0,3.0]), it will not be detected as a list of numbers by your code, and not converted to a FloatVector. (You might also generally consider using numpy arrays rather than lists, as at least their data have a well-defined type.)

geoo89 commented on opencdms_process/process/rinstat/cdms_products.py:
```diff
@@ -0,0 +1,1513 @@
+                    r_params[key] = StrVector(r_params[key])
+                elif isinstance(r_params[key][0], float):
+                    r_params[key] = FloatVector(r_params[key])
+        elif isinstance(r_params[key], DataFrame):
```
What happens if my list is of length 0?
For example, in the function `windrose`, the default value for `speed_cuts` is [], so this is a legitimate use case. You cannot infer the data type from the data itself in this case; ideally, your type annotation should give the hint here, but I don't immediately see how to access that at runtime.

geoo89 commented on opencdms_process/process/rinstat/cdms_products.py:
```diff
@@ -0,0 +1,1513 @@
+    facet_by: List[str] = None,
+    n_directions: int = 12,
+    n_speeds: int = 5,
+    speed_cuts: List[float] = [],
```
The default parameter for `speed_cuts` creates an issue for me when running the tests.
After passing it through `__get_r_params`, it should yield a `FloatVector([])`.

This doesn't happen, however. It remains a list.
When I run the tests, I get
```
E               rpy2.rinterface_lib.embedded.RRuntimeError: Error in (function (data, speed, direction, facet_by = NULL, n_directions = 12,  : 
E                 Assertion on 'speed_cuts' failed: Must be of type 'numeric', not 'list'.
```

Changing it to None doesn't help:
```
E               rpy2.rinterface_lib.embedded.RRuntimeError: Error in (function (data, speed, direction, facet_by = NULL, n_directions = 12,  : 
E                 Assertion on 'speed_cuts' failed: Must be of type 'numeric', not 'NULL'.
```

Explicitly setting it to `FloatVector([])` works for me, but I think the proper way is setting it to `NA_Real` (which also works), comparing it to the R version: https://search.r-project.org/CRAN/refmans/clifro/html/windrose.html

While in this case there is a solution for the default value, what if the user passes an empty list?
This seems to point to a more fundamental problem for me, though: The type annotations are not enforced (or considered) at runtime, making the function `__get_r_params` not do its job well. Is there some configuration that you used here to make the tests pass on your end?

Because many of my test errors seem to be caused by similar issues, I'll wait until we have discussed how to proceed more generally before going through them one by one.

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+        metadata=stations_niger,
+        file_path=output_file_actual,
+    )
+    assert __is_expected_file("export_cdt_actual010.csv")
```
What is the reason you're comparing files rather than dataframes here? Is it because this function specifically tests exporting to file?
This test fails for me because the expected file has windows line endings `\r\n`, my produced files have UNIX line endings `\n`. Apart from that, the outputs seem to be the same (however I could imagine them being different depending on the csv export settings, floating point rounding, etc).

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+        station="station_name",
+        facet_by="stations",
+    )
+    assert __is_expected_file(file_name_actual)
```
We already discussed this, but yes, all image file comparison tests fail; the images look similar, but the file sizes are vastly different probably due to different jpeg compression.

I would propose saving the expected images as PNG (which is loss-less, as opposed to jpeg), and then comparing them pixel by pixel. While this may not account for differences in fonts, anti-aliasing, etc used by the library, at least it has a better chance of passing on other OSes without being conceptually more complicated.

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+        day="day",
+    )
+    assert __is_expected_dataframe(
+        data=actual, file_name="inventory_table_actual010.csv"
```
This test fails for me: Where the expected file simply has blanks for `NA`, my output has `NA_character_ `

geoo89 replied:
I haven't yet figured out where in the code this originates.

geoo89 replied:
The DataFrame that is written to csv contains NA_character_ objects (<class 'rpy2.rinterface_lib.sexp.NACharacterType'>).
This gets written to CSV as "NA_character_" (by pandas.Dataframe.to_csv) rather than an empty string.

geoo89 commented on tests/test_cdms_products.py:
```diff
@@ -0,0 +1,1166 @@
+    )
+
+    # test that max and min are correctly calculated
+    actual = cdms_products.climatic_extremes(
```
This fails for me with
```
E               rpy2.rinterface_lib.embedded.RRuntimeError: Error in assert_column_names(data, by) : 
E                 Assertion on 'columns' failed: Must be of type 'character', not 'list'.
```
Haven't had the chance to investigate more deeply.

geoo89 replied:
It appears that, again, an empty list is not converted into StrVector in this case in __get_r_params, causing a type mismatch.
Test passes if I convert empty lists to StrVector in __get_r_params (but of course this is a hack, not a fix).

geoo89 reviewed (commented):
I looked at the failing tests in my system, pointed them out and for some investigated the causes and mitigation a bit. Some are related to comparing binary files, while some seem to be related to type checks that fail for me (not sure why they don't for you) -- either way, I feel like the lack of proper typing in python is causing some issues in some of the functionality. I haven't had the chance to get to the bottom of everything, maybe we can have a short discussion first.

Two other remarks:
- It might be worthwhile writing setUp/tearDown methods for the tests that create the folder for the actual output and delete it again, respectively (with the latter maybe possible to disable via a parameter, as keeping the files seems useful for debugging).
- It might be useful to make the test data generation reproducible, by providing the complete R code that generates it. Right now, there are just some snippets in test file.

faysal-ishtiaq commented:
Need to install the following packages on Ubuntu 22.04

```bash
libcurl4-openssl-dev libssl-dev libnetcdf-dev libfontconfig1-dev libharfbuzz-dev libfribidi-dev libnetcdf-dev
```

Need to install these packages in R as dependencies
```bash
devtools textshaping systemfonts ncdf4
```

Now the tests are failing with

```
========================================================================================== short test summary info ===========================================================================================
FAILED tests/test_cdms_products.py::test_climatic_extremes - rpy2.rinterface_lib.embedded.RRuntimeError: Error in assert_column_names(data, by) : 
FAILED tests/test_cdms_products.py::test_climatic_missing - FileNotFoundError: [Errno 2] No such file or directory: '/home/faysal/PycharmProjects/opencdms-process-dewit/tests/results_actual/climatic_missing_actual010.csv'
FAILED tests/test_cdms_products.py::test_climatic_summary - rpy2.rinterface_lib.embedded.RRuntimeError: Error in assert_column_names(data, by) : 
FAILED tests/test_cdms_products.py::test_export_cdt - rpy2.rinterface_lib.embedded.RRuntimeError: Error in file(file, ifelse(append, "a", "w")) : 
FAILED tests/test_cdms_products.py::test_export_cdt_daily - rpy2.rinterface_lib.embedded.RRuntimeError: Error in file(file, ifelse(append, "a", "w")) : 
FAILED tests/test_cdms_products.py::test_export_cdt_dekad - rpy2.rinterface_lib.embedded.RRuntimeError: Error in file(file, ifelse(append, "a", "w")) : 
FAILED tests/test_cdms_products.py::test_export_climdex - rpy2.rinterface_lib.embedded.RRuntimeError: Error in file(file, ifelse(append, "a", "w")) : 
FAILED tests/test_cdms_products.py::test_export_geoclim - rpy2.rinterface_lib.embedded.RRuntimeError: Error in file(file, ifelse(append, "a", "w")) : 
FAILED tests/test_cdms_products.py::test_export_geoclim_month - rpy2.rinterface_lib.embedded.RRuntimeError: Error in file(file, ifelse(append, "a", "w")) : 
FAILED tests/test_cdms_products.py::test_histogram_plot - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_inventory_plot - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_inventory_table - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_timeseries_plot - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_windrose - rpy2.rinterface_lib.embedded.RRuntimeError: Error in (function (data, speed, direction, facet_by = NULL, n_directions = 12,  : 
```

lloyddewit commented:
@faysal-ishtiaq Thank you for reporting this.
I just rebuilt and ran all the test_cdms_products tests. On my machine, all tests passed successfully without errors.
The errors you reported are most likely due to differences between our Ubuntu, R or Python environments.
How urgent is this issue for you?

lloyddewit commented:
@faysal-ishtiaq 
From the error messages, it looks as if the error is happening when executing the R code.
Do you have much experience with R? If so, do you get the same errors if you call the R code directly (e.g. from within RStudio)?

faysal-ishtiaq commented:
I do not have experience with R. But I will try running the R code directly and update you on that.

faysal-ishtiaq commented:
I can confirm that there is nothing wrong with the R installation. The simplest but most sensible proof would be, "while there were errors in tests, there also were passing tests".

However, I have made some slight changes to the code and now almost all tests are passing without failure or error.

Here is the diff between your last commit and my last commit: https://github.com/faysal-ishtiaq/opencdms-process/compare/bb65b64c710ac5878849866c426373fc1c213aa8...3e4b7a2026b29bff1f7ca53215e91842eb5a8215

Please note that there are some formatting-related changes in the code. And I also removed some comments from the tests.

Anyway, here is the current status of tests

```python
========================================================================================== short test summary info ===========================================================================================
FAILED tests/test_cdms_products.py::test_histogram_plot - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_inventory_plot - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_inventory_table - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_timeseries_plot - AssertionError: assert False
FAILED tests/test_cdms_products.py::test_windrose - AssertionError: assert False
================================================================================= 5 failed, 13 passed, 38 warnings in 21.03s =================================================================================
```

As you have already noticed, these tests are plotting data and checking that plot against pre-saved plots. What seems interesting to me is, the pre-saved plots and generated plots look similar, but they vary in size. Probably that is why they are not same to the `filecmp.cmp` function. Can we separate the side-effect (saving the plot to disk) from the original function and in unit-test, can we compare the function outputs rather than `jpg` or `png` files?

lloyddewit commented:
@faysal-ishtiaq Excellent that you found a solution and the tests are now all running - thank you!

> I can confirm that there is nothing wrong with the R installation. The simplest but most sensible proof would be, "while there were errors in tests, there also were passing tests".

I agree that if at least some tests run, then your R environment is installed and functional.
I was concerned that our environments may be slightly different. But if all tests are now running, and the tests that fail have graphs that are visually correct, then I think we can safely assume that your R environment is OK.
 
> However, I have made some slight changes to the code and now almost all tests are passing without failure or error.

Fantastic!

> Here is the diff between your last commit and my last commit: [faysal-ishtiaq/opencdms-process@bb65b64...3e4b7a2](https://github.com/faysal-ishtiaq/opencdms-process/compare/bb65b64c710ac5878849866c426373fc1c213aa8...3e4b7a2026b29bff1f7ca53215e91842eb5a8215)

- For the array parameter default values, it's interesting that you needed to replace empty arrays with `None`. I wonder if this is related to the Python version. I am using Python 3.8, which version are you using?
- line 287 - this line seems to be reversing the change on line 230, is this needed?
- line 1387 - the value seems incompatible with the type. I know Python doesn't check this. Would 'None' or another value be a cleaner solution?
- Package version numbers - I noticed that you upgraded to newer versions of some pacakages. If these work, then I think this is good.
- other changes look good to me.
- It would be good to review/merge your changes via the standard process. Should we merge this PR and then push your changes to a new PR?

> Please note that there are some formatting-related changes in the code. And I also removed some comments from the tests.

No problem. I use VS Code and the 'Black' formatter. Do you use a different automatic formatter?

> Anyway, here is the current status of tests
> As you have already noticed, these tests are plotting data and checking that plot against pre-saved plots. What seems interesting to me is, the pre-saved plots and generated plots look similar, but they vary in size. Probably that is why they are not same to the `filecmp.cmp` function. 

Yes, I also found that files generated directly in R, and files generated via rpy2 are not binary compatible. I tried to specify the file configuration as strictly as possible to try and create binary compatibility but it seems your/mine output is not binary compatible.
As a short-term solution, you could visually compare the graphs. If they look the same, you could move your output files to the expected output folder. Your tests should then all pass and, in future, correctly detect any local regression.

> Can we separate the side-effect (saving the plot to disk) from the original function and in unit-test, can we compare the function outputs rather than `jpg` or `png` files?

I don't know an easy way to do this. This may be a nice thing to attempt in the longer term.

faysal-ishtiaq commented:
@lloyddewit 

- I have changed the default values to immutable values so that we don't face bugs like this https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/
- And then later I handled them in a way that they are compatible with R runtime (I might mess things up as I do not have experience with R)
- I switched to latest versions of dependencies just to make sure that I am not facing any old bug/incompatibility
- I used black this time so that the diffs are minimal.

Here is the latest diff: https://github.com/faysal-ishtiaq/opencdms-process/compare/bb65b64c710ac5878849866c426373fc1c213aa8...cb150f7aa8282a50d85c2a7395a6f54a0716cd85

lloyddewit commented:
@faysal-ishtiaq Thank you for the helpful comments.

>     * I have changed the default values to immutable values so that we don't face bugs like this https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

Interesting link, I learned something. 
The first time we apply this, could we add this link in the code comment to prevent another developer inadvertently 'optimising' the code?

>     * I switched to latest versions of dependencies just to make sure that I am not facing any old bug/incompatibility

great

>     * I used black this time so that the diffs are minimal.

thank you 

> Here is the latest diff: [faysal-ishtiaq/opencdms-process@bb65b64...cb150f7](https://github.com/faysal-ishtiaq/opencdms-process/compare/bb65b64c710ac5878849866c426373fc1c213aa8...cb150f7aa8282a50d85c2a7395a6f54a0716cd85)

can we merge this PR first and then you add your diffs as a new PR?

> Should we probably skip testing the plots as they are not useful at the moment anyway?

If you comment out these tests with a comment, then no problem from my side.

isedwards commented:
> Another option is to push the changes to Stephen's branch. Then when this PR is merged, it has all the changes already.

Thanks @dannyparsons. I'm not certain which branch Faysal's latest commits are on.

@faysal-ishtiaq could you send Stephen the `git pull` command he would need to bring your changes onto his branch?

faysal-ishtiaq commented:
Sure. I will create the PR tonight.
